### PR TITLE
#546 'View Bout' action in Gmail

### LINKED ISSUE #546: "View Bout" action in Gmail
Let's add support for subject-line links in Gmail to the email subscription messages. To be clear, I'm referring to those links you see at the far right of the subject line. For example, these Github email notifications:

image

Let's make it similar to how Github does it - it should say "View Bout" and should take the receiver of the email directly to the bout being referred to.

As far as I know we have many Gmail users and this is a minor but beneficial feature for them.

See:
https://github.com/blog/1891-view-issue-pull-request-buttons-for-gmail
https://developers.google.com/gmail/markup/reference/go-to-action#view_action

### PR DESCRIPTION
**Pull request by mbarbieri**
Added HTML code to generate the ViewAction in Gmail as per documentation:
https://developers.google.com/gmail/markup/reference/go-to-action#view_action

Tests performed only by manually printing the mail message.
We need a way to test real mail, or if already there a proper documentation on how to do it.
I will raise an issue on that.
Fixes #546

### CODE REVIEW

**Comment by karato** [comment]
@mbarbieri I will find a reviewer for your pull requests shortly, thanks for contribution!

**Comment by karato** [comment]
@caarlos0 do this review

**Comment by caarlos0** [comment]
@mbarbieri please write some tests.

**Comment by mbarbieri** [comment]
@caarlos0 done

File: netbout-web/src/main/java/com/netbout/email/GmailViewAction.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            .append("  <meta itemprop=\"description\" content=\"View Bout\"/>")
+            .append("  <div itemprop=\"potentialAction\" itemscope itemtype=\"http://schema.org/ViewAction\">")
+            .append("    <link itemprop=\"target\" href=\"http://www.netbout.com/b/%d\"/>")
+            .append("    <meta itemprop=\"name\" content=\"View Bout\"/>")
+            .append("  </div>")
+            .append(" <div itemprop=\"publisher\" itemscope itemtype=\"http://schema.org/Organization\">")
+            .append("  <meta itemprop=\"name\" content=\"Netbout\"/>")
+            .append("  <link itemprop=\"url\" href=\"http://www.netbout.com/\"/>")
+            .append(" </div>")
+            .append("</div>");
```
Comment (caarlos0) [inline netbout-web/src/main/java/com/netbout/email/GmailViewAction.java]: @mbarbieri what you think about put this content in a separete HTML file and load it here?

**Comment by caarlos0** [comment]
@mbarbieri one more comment

File: netbout-web/src/main/java/com/netbout/email/GmailViewAction.java (hunk 1 above)
Comment (mbarbieri) [inline netbout-web/src/main/java/com/netbout/email/GmailViewAction.java]: @caarlos0 I thought about it. 
Maybe we can have mail templates using velocity or similar, but I thought it was too much for such a small thing. Agree that these rows are not the best code... If only Java had multiline strings...

**Comment by mbarbieri** [comment]
@caarlos0 one comment

File: netbout-web/src/main/java/com/netbout/email/GmailViewAction.java (hunk 1 above)
Comment (caarlos0) [inline netbout-web/src/main/java/com/netbout/email/GmailViewAction.java]: @mbarbieri yeah... well, let's leave it as it is for now... thanks

**Comment by caarlos0** [comment]
@rultor merge

File: netbout-web/src/main/java/com/netbout/email/GmailViewAction.java (hunk 1 above)
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/email/GmailViewAction.java]: @mbarbieri @caarlos0 I think we would use [Xembly](https://github.com/yegor256/xembly) in this case. We already have the Xembly dependencies in the project
like this
```
String xml = new Xembler(
  new Directives()
    .add("root")
    .add("order")
    .attr("id", "553")
    .set("$140.00")
).xml();
```

**Comment by dmzaytsev** [comment]
@mbarbieri one comment from me.
Also I think we need real test for it. Please take a look `EmAliasITCase.sendsRealEmail()`. You could add a new test into `EmMessagesITCase.java` then send real message check it and mark this test `@Ignore`
The mail sending is an important part of the project. I don't want a mistake here

File: netbout-web/src/main/java/com/netbout/email/GmailViewAction.java (hunk 1 above)
Comment (caarlos0) [inline netbout-web/src/main/java/com/netbout/email/GmailViewAction.java]: @dmzaytsev indeed, I forgot about this...

**Comment by caarlos0** [comment]
@dmzaytsev good point.

**Comment by caarlos0** [comment]
@mbarbieri you also need a merge here

**Comment by mbarbieri** [comment]
@dmzaytsev migrated to Xembly.
This is what I get when running ```EmAliasITCase.sendsRealEmail()``` test:
```
---------- MESSAGE FOLLOWS ----------
From: alias19082357 <no-reply@netbout.com>
To: alias298950682 <yegor@teamed.io>
Message-ID: <169521511.1.1433365859902.JavaMail.matteo@pc>
Subject: #1: untitled
MIME-Version: 1.0
Content-Type: multipart/alternative; 
    boundary="----=_Part_0_576507622.1433365859750"
X-Peer: 127.0.0.1

------=_Part_0_576507622.1433365859750
Content-Type: text/html;charset="utf-8"
Content-Transfer-Encoding: 7bit

<p>how are you?</p>

<p>--<br/>to reply click here: 
http://www.netbout.com/b/1</p>
</p>
<?xml version="1.0" encoding="UTF-8"?>
<div itemscope="" itemtype="http://schema.org/EmailMessage">
   <meta content="View Bout" itemprop="description"/>
   <div itemprop="potentialAction" itemscope="" itemtype="http://schema.org/ViewAction">
      <link href="http://www.netbout.com/b/1" itemprop="target"/>
      <meta content="View Bout" itemprop="name"/>
   </div>
   <div itemprop="publisher" itemscope="" itemtype="http://schema.org/Organization">
      <meta content="Netbout" itemprop="name"/>
      <link href="http://www.netbout.com" itemprop="url"/>
   </div>
</div>
------=_Part_0_576507622.1433365859750--
------------ END MESSAGE ------------
```
I'm still not able to run tests and send real mails.

**Comment by dmzaytsev** [comment]
@mbarbieri thats real email from github, it's different, I see ` <link ..  itemprop="target"/>` and `<link itemprop="url"`. I don't know maybe your email will display properly. Therefore I suggested to send the real email. Why you don't be able do it? I don't say run in the tests, You have to do it once and see how gmail will display your message. please see `EmMessagesITCase.java`

```
<div itemscope itemtype="http://schema.org/EmailMessage">
  <div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
    <link itemprop="url" href="https://github.com/yegor256/netbout/pull/659#issuecomment-108618825"></link>
    <meta itemprop="name" content="View Pull Request"></meta>
  </div>
  <meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>
```

**Comment by mbarbieri** [comment]
@dmzaytsev the code generated is valid as per [Google Verifier](https://www.google.com/webmasters/markup-tester/)
The problem is that in order to show the link in Gmail we have to register to Google ([read here](https://developers.google.com/gmail/markup/registering-with-google)) and mail must be authenticated via DKIM or SPF.
So my changes will do no harm but won't work until we register to Google and sign mails.

**Comment by dmzaytsev** [comment]
@mbarbieri I checked your sample email, it's not valid
As I written above there is problem with `<link .. itemprop="target"/>` /  `<link itemprop="url".`

image

**Comment by dmzaytsev** [comment]
@yegor256 We have DKIM-Signature in the netbout messages already, right?

**Comment by dmzaytsev** [comment]
@yegor256 maybe you already [registered](https://developers.google.com/gmail/markup/registering-with-google) on Google?

**Comment by yegor256** [comment]
@dmzaytsev no, I didn't do anything like that

**Comment by yegor256** [comment]
@dmzaytsev I don't know what DKIM is...

**Comment by dmzaytsev** [comment]
@yegor256 I see the valid DKIM signature  in the messages from netbout.

```
Authentication-Results: mx.google.com;
       spf=neutral (google.com: 216.22.15.79 is neither permitted nor denied by best guess record for domain of no-reply@netbout.com) smtp.mail=no-reply@netbout.com;
       dkim=pass header.i=@smtpcorp.com
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=smtpcorp.com; s=a0-2; h=Feedback-ID:X-Smtpcorp-Track:Date:Subject:
	Message-ID:To:From; bh=7hozHk9pU1lh2XofeFlaTR3lpE+LI1T89r2cAyAmqfI=; b=0LUoR/
	nhWBRh8+o2i4MHyPX4JyjwAhS7eUgFixAjRNcUxZHs/4DInCXedLIY5nUAiII1EAbmbD+g43bhgYM
	p0Zy9LQHTwf1/1KR5ax00TU4vBnZ+/c0qiLhRQvalN6t43akowScO6xepHF1t3WCivMO1JPj1q9gp
	TFw6RhUBNVy2g4SkcsOAxm+hHXd+vy6s5f6H5GnCvpMaqpWoxfzBPZjkvq3hhz43TKnrVj7CoOw4a
	iLASYgSCnKXesGEqWND0XE8dlwtA+UCgGUwq7DoPkVr3pbPJNV+5lXimjkhIoko6Ufc5LnHRjjTOR
	+25oLLegJcSiDpGyUNJ7yYYmP7iA==;
From: longtimeago <no-reply@netbout.com>
To: dmzaytsev <dmitry.zaytsev@gmail.com>
Message-ID: <1670670550.56.1433451008982.JavaMail.u3771@48d01e9c-0c7c-4f8d-a3b1-0982c2e77d69>
```

so we need to fill out the registration form only
https://developers.google.com/gmail/markup/registering-with-google

Could you help us?
Or maybe I should create a ticket for this?

**Comment by yegor256** [comment]
@dmzaytsev I'm ready to fill the form, but I need to send an email to them first (to `schema.whitelisting+sample@gmail.com`). And this test email must contain a schema markup. Can you give me that email?

**Comment by dmzaytsev** [comment]
@yegor256 they say we have to send the real email from the production server...
We can merge this request and get any message from netbout.
But @mbarbieri should fix the issue with Email Markup Tester before

**Comment by yegor256** [comment]
@dmzaytsev but our production server doesn't send emails to their address. I think I should send it myself, using our production SMTP credentials. I will send it from `EmMessageITCase`. But I need a proper message body for it, with proper "schema" inside

**Comment by dmzaytsev** [comment]
@yegor256 this is proper email. I checked it on https://www.google.com/webmasters/markup-tester/
```
---------- MESSAGE FOLLOWS ----------
From: alias19082357 <no-reply@netbout.com>
To: alias298950682 <yegor@teamed.io>
Message-ID: <169521511.1.1433365859902.JavaMail.matteo@pc>
Subject: #1: untitled
MIME-Version: 1.0
Content-Type: multipart/alternative; 
    boundary="----=_Part_0_576507622.1433365859750"
X-Peer: 127.0.0.1

------=_Part_0_576507622.1433365859750
Content-Type: text/html;charset="utf-8"
Content-Transfer-Encoding: 7bit

<p>how are you?</p>

<p>--<br/>to reply click here: 
http://www.netbout.com/b/1</p>
</p>
<?xml version="1.0" encoding="UTF-8"?>
<div itemscope="" itemtype="http://schema.org/EmailMessage">
   <meta content="View Bout" itemprop="description"/>
   <div itemprop="potentialAction" itemscope="" itemtype="http://schema.org/ViewAction">
      <link href="http://www.netbout.com/b/1" itemprop="url"/>
      <meta content="View Bout" itemprop="name"/>
   </div>
   <div itemprop="publisher" itemscope="" itemtype="http://schema.org/Organization">
      <meta content="Netbout" itemprop="name"/>
      <link href="http://www.netbout.com" itemprop="url"/>
   </div>
</div>
------=_Part_0_576507622.1433365859750--
------------ END MESSAGE ------------
```

**Comment by mbarbieri** [comment]
@dmzaytsev if you look here
https://developers.google.com/gmail/markup/reference/go-to-action#view_action
the `potentialAction` doesn't have an `url`, but only a `target`...
I added both and it passed.

**Comment by mbarbieri** [comment]
@dmzaytsev I'll fix later

**Comment by mbarbieri** [comment]
@dmzaytsev @yegor256 now the generated body is valid
```
<p>Are you using GMail?</p>

<p>--<br/>to reply click here: 
http://www.netbout.com/b/1
</p><p style="color:#C8C8C8;font-size:2px;">
83229249095092</p>
<?xml version="1.0" encoding="UTF-8"?>
<div itemscope="" itemtype="http://schema.org/EmailMessage">
   <meta content="View Bout" itemprop="description"/>
   <div itemprop="potentialAction" itemscope="" itemtype="http://schema.org/ViewAction">
      <link href="http://www.netbout.com/b/1" itemprop="target"/>
      <link href="http://www.netbout.com/b/1" itemprop="url"/>
      <meta content="View Bout" itemprop="name"/>
   </div>
   <div itemprop="publisher" itemscope="" itemtype="http://schema.org/Organization">
      <meta content="Netbout" itemprop="name"/>
      <link href="http://www.netbout.com" itemprop="url"/>
   </div>
</div>
```

**Comment by dmzaytsev** [comment]
@mbarbieri thanks!
@yegor256 if you didn't send yet please use the message body as above

**Comment by dmzaytsev** [comment]
@mbarbieri The Travis build failed

**Comment by mbarbieri** [comment]
@dmzaytsev fixed.

**Comment by dmzaytsev** [comment]
@caarlos0 are we ready to merge?

**Comment by dmzaytsev** [comment]
@mbarbieri thanks!

**Comment by dmzaytsev** [comment]
@caarlos0 ping

**Comment by caarlos0** [comment]
@rultor merge

**Comment by caarlos0** [comment]
@dmzaytsev sorry the delay!

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by karato** [comment]
@elenavolokhova please, let us know what do you think about this ticket, according to our [QA rules](http://www.teamed.io/qa.html)

**Comment by karato** [comment]
@rultor please deploy

**Comment by elenavolokhova** [comment]
@caarlos0 
According to our [quality rules](http://www.teamed.io/qa.html):
 > The code reviewer found at least three problems in the code.
 
Two issues were found during review.
Please confirm that you will try to find at least three problems next time or ask to reassign CR to another reviewer.

**Comment by caarlos0** [comment]
@elenavolokhova sure, Ill try

**Comment by elenavolokhova** [comment]
@karato Quality is acceptable

**Comment by karato** [comment]
> @karato Quality is acceptable

@elenavolokhova thanks for the review, we'll try better next time

**Comment by karato** [comment]
@caarlos0 10 mins was added to the account of @elenavolokhova (for QA review), in transaction 58968674. thank you, added **23 mins** to your acc, payment num is `AP-9P365437X5342191G`, 144 hours and 54 mins spent in total to complete. you're getting extra minutes here (c=8). added +23 to your rating, now it is equal to [+2279](http://www.netbout.com/b/35095?open=rating)

**Pull request merged**